// hono-sessions/script/src/store/CookieStore.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const deps_js_1 = require("../../deps.js");
const mod_js_1 = require("../../mod.js");
/**
 * Cookie storage driver class
 */
class CookieStore {
    constructor(options) {
        Object.defineProperty(this, "encryptionKey", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "cookieOptions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "sessionCookieName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.encryptionKey = options?.encryptionKey;
        this.cookieOptions = options?.cookieOptions;
        this.sessionCookieName = options?.sessionCookieName || 'session';
    }
    async getSession(c) {
        let session_data_raw;
        const sessionCookie = (0, deps_js_1.getCookie)(c, this.sessionCookieName);
        if (this.encryptionKey && sessionCookie) {
            // Decrypt cookie string. If decryption fails, return null
            try {
                session_data_raw = (await (0, mod_js_1.decrypt)(this.encryptionKey, sessionCookie));
            }
            catch {
                return null;
            }
            // Parse session object from cookie string and return result. If fails, return null
            try {
                const session_data = JSON.parse(session_data_raw);
                return session_data;
            }
            catch {
                return null;
            }
        }
        else {
            return null;
        }
    }
    async createSession(c, initial_data) {
        const stringified_data = JSON.stringify(initial_data);
        (0, deps_js_1.setCookie)(c, this.sessionCookieName, this.encryptionKey ? await (0, mod_js_1.encrypt)(this.encryptionKey, stringified_data) : stringified_data, this.cookieOptions);
    }
    async deleteSession(c) {
        (0, deps_js_1.setCookie)(c, this.sessionCookieName, this.encryptionKey ? await (0, mod_js_1.encrypt)(this.encryptionKey, '') : '', this.cookieOptions);
    }
    async persistSessionData(c, session_data) {
        const stringified_data = JSON.stringify(session_data);
        (0, deps_js_1.setCookie)(c, this.sessionCookieName, this.encryptionKey ? await (0, mod_js_1.encrypt)(this.encryptionKey, stringified_data) : stringified_data, this.cookieOptions);
    }
}
exports.default = CookieStore;
