// hono-sessions/esm/src/store/cloudflare/CloudflareD1Store.js
export class CloudflareD1Store {
    constructor(tableName = 'sessions') {
        Object.defineProperty(this, "db", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "tableName", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.tableName = tableName;
    }
    async getSessionById(sessionId) {
        const session = await this.db.prepare(`SELECT data FROM ${this.tableName} WHERE id = ?`)
            .bind(sessionId)
            .first('data');
        if (session) {
            return JSON.parse(session);
        }
        else {
            return null;
        }
    }
    async createSession(sessionId, initialData) {
        await this.db.prepare(`INSERT INTO ${this.tableName} (id, data) VALUES (?, ?)`).bind(sessionId, JSON.stringify(initialData)).run();
    }
    async deleteSession(sessionId) {
        await this.db.prepare(`DELETE FROM ${this.tableName} WHERE id = ?`).bind(sessionId).run();
    }
    async persistSessionData(sessionId, sessionData) {
        await this.db.prepare(`UPDATE ${this.tableName} SET data = ? WHERE id = ?`).bind(JSON.stringify(sessionData), sessionId).run();
    }
}
